### PR TITLE
Support for data volume IOPS + snapshot ID

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area storage
/kind enhancement
/priority normal
/platform aws

**What this PR does / why we need it**:
This PR adds support for configuring the IOPS for all EBS disks attached to shoot worker nodes (earlier, it was only possible to configure it for the root disk). Also, it adds support for providing a snapshot ID for data volume that is used to restore the data disk with the content stored in the snapshot. Moreover, a rolling update will now be triggered when the data volume configuration gets changed, similar to how a rolling update gets triggered when the root volume configuration gets changed.

**Which issue(s) this PR fixes**:
Ref gardener/gardener#2354

**Special notes for your reviewer**:
✅  Depends on `v0.30.0` of [gardener/machine-controller-manager](https://github.com/gardener/machine-controller-manager) that supports the `snapshotID` field in the `AWSMachineClass` (used as part of #116).

**Release note**:

```noteworthy user
It is now possible to configure the IOPS for additional data volumes (earlier, it was only possible to configure them for the root volume). Additionally, it is possible to provide a EBS snapshot ID for a data volume whose contents will be restored to this volume. Please consult [this documentation](https://github.com/gardener/gardener-extension-provider-aws/blob/master/docs/usage-as-end-user.md#workerconfig) for more information.
```

### CODE REVIEW

File: docs/usage-as-end-user.md [hunk 1]
```diff
@@ -160,8 +143,31 @@ spec:
...
+  iops: 12345
+  snapshotID: snap-1234
+```
+
+The `.volume.iops` is the number of I/O operations per second (IOPS) that the volume supports.
+For `io1` volume type, this represents the number of IOPS that are provisioned for the volume.
+For `gp2` volume type, this represents the baseline performance of the volume and the rate at which the volume accumulates I/O credits for bursting. For more information about General Purpose SSD baseline performance, I/O credits, and bursting, see Amazon EBS Volume Types (http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSVolumeTypes.html) in the Amazon Elastic Compute Cloud User Guide.\
+Constraint: Range is 100-20000 IOPS for `io1` volumes and 100-10000 IOPS for `gp2` volumes.
+
+The `.dataVolumes` can optionally contain configurations for the data volumes stated in the `Shoot` specification.
```
Comment (danielfoehrKn) [inline docs/usage-as-end-user.md:167]:
```suggestion
The `.dataVolumes` can optionally contain configurations for the data volumes stated in the `worker.dataVolumes`  specification in the `Shoot`.
```

File: pkg/controller/worker/machines.go [hunk 2]
```diff
@@ -286,3 +294,39 @@ func computeEBSDeviceNameForIndex(index int) (string, error) {
...
+
+	for _, dv := range workerConfig.DataVolumes {
+		if dv.IOPS != nil {
+			additionalData = append(additionalData, strconv.FormatInt(*dv.IOPS, 10))
+		}
+
+		if dv.SnapshotID != nil {
+			additionalData = append(additionalData, *dv.SnapshotID)
+		}
+	}
```
Comment (tim-ebert) [inline pkg/controller/worker/machines.go:329]: Hmmm, the worker provider config is already considered in `worker.WorkerPoolHash`. Is it needed to also add it to the additional hash data?

File: pkg/apis/aws/validation/shoot.go [hunk 3]
```diff
@@ -37,45 +37,51 @@ func ValidateNetworking(networking core.Networking, fldPath *field.Path) field.E
...
+			dataVolPath := fldPath.Child("dataVolumes").Index(j)
+
+			if volume.Name == nil {
+				allErrs = append(allErrs, field.Required(dataVolPath.Child("name"), "must not be empty"))
+			}
+
+			allErrs = append(allErrs, validateVolume(volume.DeepCopy(), dataVolPath)...)
+
+			if volume.Type != nil && *volume.Type == string(apisaws.VolumeTypeIO1) && worker.ProviderConfig == nil {
+				allErrs = append(allErrs, field.Required(fldPath.Child("providerConfig"), fmt.Sprintf("WorkerConfig must be set if data volume type is %s (%s)", apisaws.VolumeTypeIO1, dataVolPath.Child("type"))))
```
[the remark this answers is not in the document]
Comment (rfranzke) [reply pkg/apis/aws/validation/shoot.go:72]: Why do you suggest using the `name` field? The check is about the `type`, that's why I added the path to this field.

File: pkg/controller/worker/machines.go (hunk 2 above)
Comment (rfranzke) [reply pkg/controller/worker/machines.go:329]: Yes, indeed, nice catch. Removed the code, thanks.

File: pkg/controller/worker/machines.go [hunk 4]
```diff
@@ -114,7 +114,14 @@ func (w *workerDelegate) generateMachineConfig(ctx context.Context) error {
 	for _, pool := range w.worker.Spec.Pools {
 		zoneLen := int32(len(pool.Zones))
 
-		workerPoolHash, err := worker.WorkerPoolHash(pool, w.cluster)
+		workerConfig := &awsapi.WorkerConfig{}
+		if pool.ProviderConfig != nil && pool.ProviderConfig.Raw != nil {
+			if _, _, err := w.Decoder().Decode(pool.ProviderConfig.Raw, nil, workerConfig); err != nil {
```
[the remark this answers is not in the document]
Comment (rfranzke) [reply pkg/controller/worker/machines.go:119]: When I'm doing this (also for the `InfrastructureStatus` decode a few lines above) then I get issues in the test:

```
• Failure [0.003 seconds]
Machines
/Users/rfranzke/go/src/github.com/gardener/gardener-extension-provider-aws/pkg/controller/worker/machines_test.go:53
  workerDelegate
  /Users/rfranzke/go/src/github.com/gardener/gardener-extension-provider-aws/pkg/controller/worker/machines_test.go:71
    #GenerateMachineDeployments, #DeployMachineClasses
    /Users/rfranzke/go/src/github.com/gardener/gardener-extension-provider-aws/pkg/controller/worker/machines_test.go:86
      machine images
      /Users/rfranzke/go/src/github.com/gardener/gardener-extension-provider-aws/pkg/controller/worker/machines_test.go:409
        should return the expected machine deployments for profile image types [It]
        /Users/rfranzke/go/src/github.com/gardener/gardener-extension-provider-aws/pkg/controller/worker/machines_test.go:582

        Unexpected error:
            <*errors.errorString | 0xc000d38ac0>: {
                s: "no instance profile with purpose \"nodes\" found",
            }
            no instance profile with purpose "nodes" found
        occurred

        /Users/rfranzke/go/src/github.com/gardener/gardener-extension-provider-aws/pkg/controller/worker/machines_test.go:601
------------------------------
SSSSSSSS
```

Note that this happens even after adding

```golang
								TypeMeta: metav1.TypeMeta{
									APIVersion: apiv1alpha1.SchemeGroupVersion.String(),
									Kind:       "InfrastructureStatus",
								},
```

to all relevant usages...

File: pkg/apis/aws/validation/shoot.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (tim-ebert) [reply pkg/apis/aws/validation/shoot.go:72]: Ah yes, sure. I was confused, forget it :)

File: pkg/controller/worker/machines.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (timuthy) [reply pkg/controller/worker/machines.go:119]: Oh okay, thanks for the info. Then lets proceed w/o this util function and do it later in a separate PR. The "correct" decoding is already done during validation, so we should be on the safe side.

**Pull request merged**